extensions/amp-ad-network-adsense-impl/0.1/adsense-shared-state.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export class AdsenseSharedState {
+
+  constructor() {
+
+    /** @private {!Array<string>} */
+    this.prevFmts_ = [];
+
+    /** @private {!Object<string, number>} */
+    this.slotFormatIndexMap_ = {};
+
+    /** @private {number} */
+    this.formatIndex_ = 0;
+
+    /** @private {!Object<string, number>} */
+    this.pv_ = {};
+  }
+
+  /**
+   * Returns the formats of the previous slots as a comma separated list.
+   * @param {string} format The format to add.
+   * @param {string} id Unique identifier for slot.
+   * @return {string}
+   */
+  updateAndGetPrevFmts(format, id) {
+    // The return value.
+    const prevFmts = this.prevFmts_.join(',');
+    // Associate the insertion index with the given id for future removal.
+    this.slotFormatIndexMap_[id] = this.formatIndex_++;
```

Comment:
You could remove the need for this mapping and formatIndex by storing an array of objects instead of just strings in prevFmts;

`/** @private {!Array<!{{id: string, format: string, client: string}}>} */
this.previousSlots_ = [];

/**
 * @param {string} format
 * @param {string} id
 * @param {string} client
 * @return !{{prevFmts: string, pv: number}}
 */
addNewSlot(format, id, client) {
   let result = {pv: 1, prevFmts: ''};
   this.previousSlots_.forEach(slot => {
       dev().assert(slot.id != id);
       result.prevFmts += (result.prevFmts ? '' : ',') + slot.format;
       if (slot.client == client) {
          result.pv = 2;
       }
   });
   this.previousSlots_.push({id: id, format: format, client: client});
   return result;
}

removeSlot(id) {
   this.previousSlots_ = this.previousSlots_.filter(slot => {
       return slot.id != id;
   });
}`